### PR TITLE
Feature/jzy 254 improve callbacks

### LINKED ISSUE #254: [EmulGL] GL functions to support callbacks as Callable
Functions like `glutCreateMenu` that accepts callback function name as an argument could have overloaded counterparts that accept lambda as an argument.
I.e. instead of 
`public void glutCreateMenu(String func)` one could have `public void glutCreateMenu(Callable func)` - this would increase performance (of course marginally) and code readability due to IDE code navigation support

### PR DESCRIPTION
**Pull request by valb3r**
Fixes #254 

Overview of fixes:
1. Kept old string-based functions and marked them as Deprecated
2. Introduced new lambda-based callbacks as counterparts of string based functions
3. Introduced GLUTListener which itself can replace callbacks if needed, can be registered with `glutSetListener`

### CODE REVIEW

File: jzy3d-jGL/jzy3d-jGL-awt/src/main/java/jgl/wt/awt/listener/callbacks/KeyboardCallback.java [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+  /**
+   * Sets the keyboard callback for the current window.
+   * void glutKeyboardFunc (void (*func)(unsigned char key, int x, int y))
+   *
+   * @param target Rendering target component
+   * @param key    Generated ASCII character (i.e. 'a' or 'A' depending on SHIFT key)
+   * @param x      Window relative coordinate x of mouse when event has occurred
+   * @param y      Window relative coordinate y of mouse when event has occurred
+   */
+  void onKeyboard(Component target, char key, int x, int y);
```
Comment (jzy3d) [inline jzy3d-jGL/jzy3d-jGL-awt/src/main/java/jgl/wt/awt/listener/callbacks/KeyboardCallback.java:17]: Why not making it public?

File: jzy3d-jGL/jzy3d-jGL-awt/src/main/java/jgl/wt/awt/listener/GLUTReflectiveCallbackListener.java [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+import java.awt.*;
+import java.lang.reflect.InvocationTargetException;
+import java.lang.reflect.Method;
+
+/**
+ * Old reflection based 'Method' callback based on method name.
+ */
+public class GLUTReflectiveCallbackListener implements GLUTListener {
+
+  private Method reshapeMethod;
```
Comment (jzy3d) [inline jzy3d-jGL/jzy3d-jGL-awt/src/main/java/jgl/wt/awt/listener/GLUTReflectiveCallbackListener.java:12]: Despite not a big deal here, I tend to always set fields protected in Jzy3D to ease overriding.

File: jzy3d-jGL/jzy3d-jGL-awt/src/main/java/jgl/wt/awt/listener/GLUTListener.java [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+   * Sets the global idle callback so that program can perform background processing tasks.
+   * void glutIdleFunc (void (*func)(void))
+   *
+   * @param target Rendering target component
+   */
+  default void onIdle(Component target) {
+  }
+
+  /**
+   * Indicates where the listener has usable {@link this#onMouse(Component, int, int, int, int)}
```
Comment (jzy3d) [inline jzy3d-jGL/jzy3d-jGL-awt/src/main/java/jgl/wt/awt/listener/GLUTListener.java:109]: Typo on **wether**?

I am surprised by this method. Is this something that appeared in GLUT, or are you following a pattern?

File: jzy3d-emul-gl-awt/src/main/java/org/jzy3d/plot3d/rendering/canvas/EmulGLCanvas.java [hunk 4]
```diff
@@ -180,12 +181,12 @@ protected void initGLUT(int width, int height) {
...
     // GLUT register doRender as display callback
 
-    myUT.glutDisplayFunc("doRender");
+    myUT.glutDisplayFunc(c -> doRender());
 
     // --------------------------------------------
     // GLUT register doReshape as resize callback
 
-    myUT.glutReshapeFunc("doReshape");
+    myUT.glutReshapeFunc((c, w, h) -> doReshape(w, h));
```
Comment (jzy3d) [inline jzy3d-emul-gl-awt/src/main/java/org/jzy3d/plot3d/rendering/canvas/EmulGLCanvas.java:189]: Keeping parameters with their complete name would make the code easier to read!

**Review by jzy3d — COMMENTED** [review]
Great contribution, thanks! I only left minor comments in the code.

File: jzy3d-jGL/jzy3d-jGL-awt/src/main/java/jgl/wt/awt/listener/callbacks/KeyboardCallback.java (hunk 1 above)
Comment (valb3r) [reply jzy3d-jGL/jzy3d-jGL-awt/src/main/java/jgl/wt/awt/listener/callbacks/KeyboardCallback.java:17]: Method in interfaces are by default public

File: jzy3d-jGL/jzy3d-jGL-awt/src/main/java/jgl/wt/awt/listener/GLUTListener.java (hunk 3 above)
Comment (valb3r) [reply jzy3d-jGL/jzy3d-jGL-awt/src/main/java/jgl/wt/awt/listener/GLUTListener.java:109]: > I am surprised by this method. Is this something that appeared in GLUT, or are you following a pattern?

There are some GLUT initialisations happening if listener supports certain capability

**Comment by jzy3d** [comment]
I ran unit and integration tests locally without seeing any regression. Manual tests show that everything works perfectly. Congratulation for this first contribution @valb3r!

**Pull request merged**